src/analysis/inner_product_space/gram_schmidt_ortho.lean

```diff
@@ -0,0 +1,342 @@
+    orthogonal_projection (𝕜 ∙ gram_schmidt 𝕜 f i) (f n) :=
+begin
+  rw gram_schmidt,
+  congr' 1,
+  exact fin.sum_univ_eq_sum_range (λ i,
+    (orthogonal_projection (𝕜 ∙ gram_schmidt 𝕜 f i) (f n) : E)) n,
+end
+
+@[simp] lemma gram_schmidt_zero (f : ℕ → E) :
+  gram_schmidt 𝕜 f 0 = f 0 := by simp [gram_schmidt]
+
+/-- Gram-Schmidt process produces an orthogonal system of vectors. -/
+theorem gram_schmidt_orthogonal' (f : ℕ → E) (a b : ℕ) (h₀ : a < b) :
+  ⟪gram_schmidt 𝕜 f a, gram_schmidt 𝕜 f b⟫ = 0 :=
+begin
+  have hc : ∃ c, b ≤ c := ⟨b + 1, by linarith⟩,
+  cases hc with c h₁,
+  induction c with c hc generalizing a b,
+  { simp only [nat.nat_zero_eq_zero, le_zero_iff] at h₁,
+    simp only [h₁, not_lt_zero'] at h₀,
+    contradiction },
+  { rw nat.le_add_one_iff at h₁,
+    cases h₁ with hb₁ hb₂,
+    { exact hc _ _ h₀ hb₁ },
+    { simp only [gram_schmidt_def 𝕜 f (c + 1), hb₂, inner_sub_right, inner_sum],
+      have h₂ : ∀ x ∈ finset.range(c + 1), x ≠ a →
+        ⟪gram_schmidt 𝕜 f a,
+          (orthogonal_projection (𝕜 ∙ (gram_schmidt 𝕜 f x)) (f (c + 1)) : E)⟫ = 0,
+      { intros x hx₁ hx₂,
+        simp only [orthogonal_projection_singleton],
+        rw inner_smul_right,
+        cases hx₂.lt_or_lt with hxa₁ hxa₂,
+        { have ha₂ : a ≤ c,
+          { rw hb₂ at h₀,
+            exact nat.lt_succ_iff.mp h₀ },
+          specialize hc x a,
+          simp only [hxa₁, ha₂, forall_true_left] at hc,
+          simp only [mul_eq_zero, div_eq_zero_iff, inner_self_eq_zero],
+          right,
+          rwa inner_eq_zero_sym at hc },
+        { rw [finset.mem_range, nat.lt_succ_iff] at hx₁,
+          specialize hc a x,
+          simp only [hxa₂, hx₁, forall_true_left] at hc,
+          simp only [mul_eq_zero, div_eq_zero_iff, inner_self_eq_zero],
+          right,
+          exact hc } },
+      rw hb₂ at h₀,
+      have ha₂ : a ∈ finset.range(c + 1) := finset.mem_range.mpr h₀,
+      rw finset.sum_eq_single_of_mem a ha₂ h₂,
+      simp only [orthogonal_projection_singleton],
+      rw inner_smul_right,
+      by_cases ⟪gram_schmidt 𝕜 f a, gram_schmidt 𝕜 f a⟫ = 0,
+      { simp only [inner_self_eq_zero] at h,
+        repeat {rw h},
+        simp only [inner_zero_left, mul_zero, sub_zero] },
+      { rw ← inner_self_eq_norm_sq_to_K,
+        simp only [h, div_mul_cancel, ne.def, not_false_iff, sub_self] } } }
+end
+
+/-- **Gram-Schmidt Orthogonalisation**
+This version is stronger than `gram_schmidt_orthogonal'` as it doesn't require `a < b` -/
+theorem gram_schmidt_orthogonal (f : ℕ → E) (a b : ℕ) (h₀ : a ≠ b) :
+  ⟪gram_schmidt 𝕜 f a, gram_schmidt 𝕜 f b⟫ = 0 :=
+begin
+  cases h₀.lt_or_lt with ha hb,
+  { exact gram_schmidt_orthogonal' 𝕜 f a b ha },
+  { rw inner_eq_zero_sym,
+    exact gram_schmidt_orthogonal' 𝕜 f b a hb }
+end
+
+open submodule set
+
+/-- Gram-Schmidt process preserves span -/
+lemma span_gram_schmidt (f : ℕ → E) (c : ℕ) :
+  span 𝕜 (gram_schmidt 𝕜 f '' Iic c) = span 𝕜 (f '' Iic c) :=
```

Comment:
Would it be easier to use `finset.range c` instead of `Iic`?
It looks a bit funny to use `Iic` because `ℕ` doesn't have a negative infinity.

Reply:
oh yes, I think that would be easier here. I will try to use range here, thank you!

Reply:
I have changed them to `range` here. Thank you!

Reply:
@jcommelin, using a finset feels strange to me here, since we're using `''` not `finset.image`. Iic seems clearest to me, and is certainly definitionally nicer.

If you're convinced finset.range is the way to go, maybe another reviewer can tiebreak

Reply:
I agree with Eric on this one -- with `Iic` we can avoid going to a `finset` and then immediately back to a `set`.

Reply:
Ok, fair enough

Reply:
@Biiiilly sorry for giving the wrong suggestion.

Reply:
that's fine, I have changed back :)